cmd/mikrotik-codegen/main.go

```diff
@@ -0,0 +1,129 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/ddelnano/terraform-provider-mikrotik/cmd/mikrotik-codegen/internal/codegen"
+)
+
+type (
+	Configuration struct {
+		SrcFile    string
+		DestFile   string
+		StructName string
+		FormatCode bool
+	}
+)
+
+func main() {
+	if err := realMain(os.Args[1:]); err != nil {
+		log.Fatalf("execution failed: %v", err)
+	}
+
+	os.Exit(0)
+}
+
+// todo:
+//
+//	omit fields by default: Id
+//	different fields for update/delete client funcs
+//		(sometimes it is Id, sometimes it's Name, etc)
+func realMain(args []string) error {
+	config, err := parseConfig(args)
+	if err != nil {
+		return err
+	}
+
+	startLine := 1
+	lineStr := os.Getenv("GOLINE")
+	if lineStr != "" {
+		lineInt, err := strconv.Atoi(lineStr)
+		if err != nil {
+			return fmt.Errorf("fail to parse GOLINE: %v", err.Error())
+		}
+		startLine = lineInt
+	}
+
+	s, err := codegen.ParseFile(config.SrcFile, startLine, config.StructName)
+	if err != nil {
+		return err
+	}
+	if s.TerraformIDField == "" {
+		return errors.New("Terraform ID field is not set to any of struct's fields")
+	}
+	if s.MikrotikIDField == "" {
+		return errors.New("Mikrotik ID field is not set to any of struct's fields")
+	}
+
+	if config.StructName == "" {
+		config.StructName = s.Name
+	}
+	// if destination was not provided via args, build it using struct name
+	// the initialization is delayed, because config.StructName might not be provided, so we get it after actual file parsing
+	// skip this not to accidently rewrite existing file
+	// if config.DestFile == "" {
+	// 	config.DestFile, err = filepath.Abs(path.Join("../mikrotik", "resource_"+utils.ToSnakeCase(config.StructName)) + ".go")
+	// 	if err != nil {
+	// 		return err
+	// 	}
+	// }
+
+	if config.DestFile == "" {
+		return errors.New("destination file must be set via flags or 'go:generate' mode must be used")
+	}
+
+	var out io.Writer
+	if config.DestFile == "-" {
+		out = os.Stdout
+	} else {
+		file, err := os.OpenFile(config.DestFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+		if err != nil {
+			return err
+		}
+		out = file
+		defer func() {
+			file.Close()
+		}()
+	}
+	writeHooks := []codegen.SourceWriteHookFunc{}
+	if config.FormatCode {
+		writeHooks = append(writeHooks, codegen.SourceFormatHook)
+	}
+	return codegen.GenerateResource(s, out, writeHooks...)
+}
+
+func parseConfig(args []string) (*Configuration, error) {
+	var (
+		destFile   = flag.String("dest", "-", "File to write result to. Default: write to stdout.")
+		srcFile    = flag.String("src", "", "Source file to parse struct from.")
+		structName = flag.String("struct", "", "Name of a struct to process.")
+		formatCode = flag.Bool("formatCode", true, "Whether to format resulting code.")
```

Comment:
I am using it for debugging.
`Go` can't format code with syntax errors, and `expecting ',', got ')' at mikrotik/resource_new.go:103:15` says nothing, because line numbers do not correspond to template line numbers 🤷

Reply:
ah, I see. Can you add that additional detail in the flag description? I didn't understand how useful the option could be (I see that debugging use case as very important).